cozy/codegen/rpython.py

```diff
@@ -0,0 +1,422 @@
+        yield
+        self.write(")")
+
+    @contextmanager
+    def stmt(self):
+        self.begin_statement()
+        yield
+        self.end_statement()
+
+    def visit_ENum(self, node):
+        self.write(str(node.val))
+
+    def visit_EBool(self, node):
+        self.write("True" if node.val else "False")
+
+    def visit_EEmptyList(self, _):
+        self.write("[]")
+
+    def visit_ESingleton(self, node):
+        self.write("[")
+        self.visit(node.e)
+        self.write("]")
+
+    def visit_EVar(self, node):
+        name = node.id
+        if name in self.instance_names:
+            self.write("self.", name)
+        else:
+            self.write(name)
+
+    def emit_unary(self, op, e):
+        self.needs[op] = True
+        self.write("_", op, "(")
+        self.visit(e)
+        self.write(")")
+
+    def visit_EUnaryOp(self, node):
+        op = node.op
+        if op in ("distinct", "reversed", "sum"):
+            self.emit_unary(op, node.e)
+        elif op == "len":
+            self.write("len(")
+            self.visit(node.e)
+            self.write(")")
+        elif op == "not":
+            self.write("not ")
+            self.visit(node.e)
+        elif op == "empty":
+            self.write("not bool(")
+            self.visit(node.e)
+            self.write(")")
+        elif op == "exists":
+            self.write("bool(")
+            self.visit(node.e)
+            self.write(")")
+        else:
+            import pdb; pdb.set_trace()
+            assert False, "Unhandled unary operation while writing RPython"
+
+    def visit_EBinOp(self, node):
+        # Rather than dispatching on op, it turns out to be simpler for us to
+        # dispatch by type!
+        if isinstance(node.type, (TInt, TBool)):
+            with self.parens():
+                self.visit(node.e1)
+                self.write(" ", node.op, " ")
+                self.visit(node.e2)
+        else:
+            import pdb; pdb.set_trace()
+            assert False, "Unhandled binary operation while writing RPython"
+
+    def visit_ECond(self, node):
+        # Quirk: Sometimes cond will be constantly true or false.
+        if isinstance(node, EBool):
+            self.visit(node.then_branch if node.val else node.else_branch)
+
+        with self.parens():
+            self.visit(node.then_branch)
+            self.write(" if ")
+            self.visit(node.cond)
+            self.write(" else ")
+            self.visit(node.else_branch)
+
+    def visit_EMap(self, node):
+        f = node.f
+        # Quirk: Sometimes we will see (map (\x -> x) xs) and we want to just
+        # have xs instead.
+        if f.arg == f.body:
+            self.visit(node.e)
+            return
+
+        self.write("[")
+        self.visit(f.body)
+        self.write(" for ")
+        # Hack: Write the lambda's arg as if it were an expr. This is close
+        # enough for now but definitely wrong. ~ C.
+        self.visit(f.arg)
```

Comment:
No, it's not as much of a hack as you think. `f.arg` is guaranteed to be an `EVar` object. If you want to be a little safer about this, just change the line to `self.write(f.arg.id)`.

Reply:
The same goes for other places you follow this pattern.

Reply:
Interesting but useful that lambdas are always single-argument. With that in mind, I'd definitely clean these spots up a bit. We are lucky that RPython supports list comprehensions.